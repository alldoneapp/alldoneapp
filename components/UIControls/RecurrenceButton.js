import React, { Component } from 'react'
import Popover from 'react-tiny-popover'
import store from '../../redux/store'
import PropTypes from 'prop-types'
import Button from './Button'
import { hideFloatPopup, showFloatPopup } from '../../redux/actions'
import RecurrenceModal from '../UIComponents/FloatModals/RecurrenceModal'
import { translate } from '../../i18n/TranslationService'
import { RECURRENCE_MAP } from '../TaskListView/Utils/TasksHelper'

class RecurrenceButton extends Component {
    constructor(props) {
        super(props)
        const storeState = store.getState()

        this.state = {
            visiblePopover: false,
            smallScreen: storeState.smallScreen,
            unsubscribe: store.subscribe(this.updateState),
        }
    }

    componentWillUnmount() {
        this.state.unsubscribe()
    }

    updateState = () => {
        const storeState = store.getState()

        this.setState({
            smallScreen: storeState.smallScreen,
        })
    }

    hidePopover = recurrence => {
        this.setState({ visiblePopover: false })
        store.dispatch(hideFloatPopup())

        if (this.props.onChangeValue !== undefined && recurrence) {
            this.props.onChangeValue(recurrence)
        }
    }

    showPopover = () => {
        this.setState({ visiblePopover: true })
        store.dispatch(showFloatPopup())
    }

    render() {
        const { task, projectId, disabled, style, inEditTask } = this.props
        const { visiblePopover, smallScreen } = this.state

        return (
            <Popover
                content={<RecurrenceModal task={task} projectId={projectId} closePopover={this.hidePopover} />}
                onClickOutside={this.hidePopover}
                isOpen={visiblePopover}
                position={['bottom', 'left', 'right', 'top']}
                padding={4}
                align={'end'}
                contentLocation={smallScreen ? null : undefined}
            >
                <Button
                    title={inEditTask && smallScreen ? null : translate(RECURRENCE_MAP[task.recurrence].large)}
                    type={'ghost'}
                    noBorder={inEditTask && smallScreen}
                    icon={'rotate-cw'}
                    buttonStyle={style}
                    onPress={this.showPopover}
                    disabled={disabled}
                />
            </Popover>
        )
    }
}

RecurrenceButton.propTypes = {
    task: PropTypes.object.isRequired,
    projectId: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    style: PropTypes.object,
    inEditTask: PropTypes.bool,
    onChangeValue: PropTypes.func,
}

RecurrenceButton.defaultProps = {
    inEditTask: false,
}

export default RecurrenceButton
