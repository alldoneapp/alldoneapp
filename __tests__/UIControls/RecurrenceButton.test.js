/**
 * @jest-environment jsdom
 */

import React from 'react'
import RecurrenceButton from '../../components/UIControls/RecurrenceButton'
import renderer from 'react-test-renderer'

const task = { id: '-Sda', name: 'My task', recurrence: { type: 'never' } }

describe('Task Recurrence Button component', () => {
    describe('Task Estimation Button snapshot test', () => {
        it('should render correctly', () => {
            const tree = renderer.create(<RecurrenceButton projectId={'-Asd'} task={task} />).toJSON()
            expect(tree).toMatchSnapshot()
        })
    })

    describe('Function updateState snapshot test', () => {
        it('should execute and render correctly', () => {
            const tree = renderer.create(<RecurrenceButton projectId={'-Asd'} task={task} />)
            expect(tree.toJSON()).toMatchSnapshot()

            tree.getInstance().updateState()
            expect(tree.toJSON()).toMatchSnapshot()
        })
    })

    describe('Function hidePopover snapshot test', () => {
        it('should execute and render correctly', () => {
            const tree = renderer.create(<RecurrenceButton projectId={'-Asd'} task={task} />)
            expect(tree.toJSON()).toMatchSnapshot()

            tree.getInstance().state.visiblePopover = true
            tree.getInstance().hidePopover()

            let state = tree.getInstance().state
            expect(state.visiblePopover).toBeFalsy()
        })
    })

    describe('Task Recurrence check unmount', () => {
        it('should unmount correctly', () => {
            const tree = renderer.create(<RecurrenceButton projectId={'-Asd'} task={task} />)
            tree.getInstance().componentWillUnmount()
        })
    })
})
