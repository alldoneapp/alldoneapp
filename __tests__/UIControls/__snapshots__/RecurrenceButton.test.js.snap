// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Recurrence Button component Function hidePopover snapshot test should execute and render correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "transparent",
      "borderColor": "#C6CDD2",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": 40,
      "justifyContent": "center",
      "maxHeight": 40,
      "minHeight": 40,
      "opacity": 1,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingVertical": 8,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#8A94A6",
          "fontFamily": "alldone",
          "fontSize": 24,
          "opacity": 1,
        }
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`Task Recurrence Button component Function updateState snapshot test should execute and render correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "transparent",
      "borderColor": "#C6CDD2",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": 40,
      "justifyContent": "center",
      "maxHeight": 40,
      "minHeight": 40,
      "opacity": 1,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingVertical": 8,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#8A94A6",
          "fontFamily": "alldone",
          "fontSize": 24,
          "opacity": 1,
        }
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`Task Recurrence Button component Function updateState snapshot test should execute and render correctly 2`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "transparent",
      "borderColor": "#C6CDD2",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": 40,
      "justifyContent": "center",
      "maxHeight": 40,
      "minHeight": 40,
      "opacity": 1,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingVertical": 8,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#8A94A6",
          "fontFamily": "alldone",
          "fontSize": 24,
          "opacity": 1,
        }
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`Task Recurrence Button component Task Points Button snapshot test should render correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "transparent",
      "borderColor": "#C6CDD2",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": 40,
      "justifyContent": "center",
      "maxHeight": 40,
      "minHeight": 40,
      "opacity": 1,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingVertical": 8,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#8A94A6",
          "fontFamily": "alldone",
          "fontSize": 24,
          "opacity": 1,
        }
      }
    >
      
    </Text>
  </View>
</View>
`;
